debian/nsexec.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH NSEXEC SECTION "February 11, 2011"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
nsexec \- program to do something
.SH SYNOPSIS
.B nsexec
.RI [ -c ] [ -g ] [mnpiuU] [-P pidfile] " program [arguments]"
.br
.SH DESCRIPTION
This manual page documents briefly the
.B nsexec
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBnsexec\fP is a program that launches commands in a set of new
namespaces.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-c
Use 'clone' rather than 'unshare' to create the namespaces.  This
is required for some namespaces, including user and pid.
.B \-m
Create a new mounts namespace
.B \-g
Launch in a new cgroup as well
.B \-n
New network namespace
.B \-u
New uts namespace (hostname)
.B \-i
New IPC namespace
.B \-U
New user namespace
.B \-p
New pid namespace
.B \-P
Name of a file into which to write the new task's pid.
.br
.SH EXAMPLES
.B nsexec -cmpnui /bin/bash
Launch a bash shell in a task with private mounts, pid, network,
uts, and IPC namespaces.
.br
.SH AUTHOR
nsexec was written by <upstream author>.
.PP
This manual page was written by Serge Hallyn <serge.hallyn@ubuntu.com>,
for the Debian project (and may be used by others).


debian/changelog
nsexec (1.22+nmu1) UNRELEASED; urgency=medium

  * Non-maintainer upload.

 -- Lars Boegild Thomsen <lbthomsen@optimal-dynamics.com>  Tue, 03 Apr 2018 22:02:13 +0800

nsexec (1.22) saucy; urgency=low

  * usernsexec: don't run newxidmap if there was no mapping to effect.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 16 Jul 2013 15:01:20 -0500

nsexec (1.21) saucy; urgency=low

  * usernsexec: drop supplemental groups

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Wed, 08 May 2013 16:46:34 -0500

nsexec (1.20saucy) saucy; urgency=low

  * rebuild for saucy

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Thu, 02 May 2013 21:19:42 -0500

nsexec (1.20) raring; urgency=low

  * update order/format of lxc.id_map entries to  match new upstream lxc
    support.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Thu, 11 Apr 2013 12:10:22 -0500

nsexec (1.19) raring; urgency=low

  * add usernsexec program

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 12 Mar 2013 11:05:00 -0500

nsexec (1.18) raring; urgency=low

  * cleaner simpler version nof usernsselfmap.
  * add options to nsexec to make it more useful for user namespaces.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Thu, 07 Mar 2013 14:40:16 -0600

nsexec (1.17) raring; urgency=low

  * add usernsselfmap program to demonstrate creating a user ns completely
    without privilege (only mapping uid 0 in container to your host uid).

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Wed, 06 Mar 2013 16:16:04 -0600

nsexec (1.16) raring; urgency=low

  * Update container-userns-convert to use new uidmapshift semantics.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Wed, 06 Mar 2013 14:56:56 -0600

nsexec (1.15) raring; urgency=low

  * Apply patch from Dwight to use nftw in uidmapshift (and other
    improvements).

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 25 Feb 2013 15:00:30 -0600

nsexec (1.14) raring; urgency=low

  * Remove debugging code accidentally left in.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 15 Jan 2013 10:40:18 -0600

nsexec (1.13) raring; urgency=low

  * uidmapshift: use lchmod in place of chmod, make less verbose, add DEBUG
    compile time option, and fix up wrong variable names.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Fri, 11 Jan 2013 20:56:07 -0600

nsexec (1.0-12) raring; urgency=low

  * build for raring

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 04 Dec 2012 21:11:17 -0600

nsexec (1.0-11) quantal; urgency=low

  * rename convert_container to container-userns-convert, and have Makefile
    install it to /usr/bin.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Wed, 31 Oct 2012 05:39:11 +0100

nsexec (1.0-10) quantal; urgency=low

  * Restore suid and sgid after chowning in uidmapshift.
  * Add a 'shift-container-uids' script to set up a container for use with
    user namespaces.
  * Drop quilt patches and switch to pure bzr format.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 30 Oct 2012 16:09:54 +0100

nsexec (1.0-9) quantal; urgency=low

  * Fix help in uidmapshift.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 29 Oct 2012 15:04:40 +0100

nsexec (1.0-8) quantal; urgency=low

  * add usernstest and uidmapshift programs.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 25 Jun 2012 16:42:50 -0500

nsexec (1.0-7) quantal; urgency=low

  * Add comments to the sample /etc/id_permissions/* files, and insert them
    through debian/rules rather than Makefile.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Thu, 10 May 2012 14:36:08 -0700

nsexec (1.0-6) quantal; urgency=low

  * install /etc/id_permission/{u,g}ids
  * Have uidmap respect /etc/id_permission/{u,g}ids
  * Make uidmap setuid-root

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Thu, 10 May 2012 13:58:18 -0700

nsexec (1.0-5) quantal; urgency=low

  * build for quantal
  * add #define for CLONE_NEWNS which appears to no longer ship in libc
    headers.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Wed, 09 May 2012 10:30:39 -0700

nsexec (1.0-4.11.10) oneiric; urgency=low

  * build for oneiric in userns-natty

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 11 Jul 2011 13:45:48 -0500

nsexec (1.0-4) natty; urgency=low

  * bugfix: add -e to getopt call (and use WEXITSTATUS).

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 11 Jul 2011 11:55:59 -0500

nsexec (1.0-3) natty; urgency=low

  * add '-e' option to return the child's return value.  By default, as
    before, return 0 as long as clone succeeded.

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Mon, 11 Jul 2011 11:52:02 -0500

nsexec (1.0-2) natty; urgency=low

  * Fix package build

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Fri, 11 Feb 2011 14:28:02 -0600

nsexec (1.0-1) natty; urgency=low

  * Initial release

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Fri, 11 Feb 2011 13:16:14 -0600


nsexec.c
/*
 * Copyright 2008 IBM Corp.
 */

#include <stdio.h>
#include <stdlib.h>
#include <sched.h>
#include <sys/syscall.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <errno.h>
#include <libgen.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <grp.h>

#include "clone.h"
int unshare(int flags);
int clone(int (*fn)(void *), void *child_stack,
        int flags, void *arg, ...
        /* pid_t *ptid, struct user_desc *tls, pid_t *ctid */ );


extern pid_t getpgid(pid_t pid);
extern pid_t getsid(pid_t pid);

static const char* procname;

static void usage(const char *name)
{
	printf("usage: %s [-h] [-c] [-mnuUip] [-P <pid-file>]"
			"[command [arg ..]]\n", name);
	printf("\n");
	printf("  -h		this message\n");
	printf("\n");
	printf("  -c		use 'clone' rather than 'unshare' system call\n");
	printf("  -g		launch in new cgroup\n");
	printf("  -m		mount namespace\n");
	printf("  -n		network namespace\n");
	printf("  -u		utsname namespace\n");
	printf("  -U		userid namespace\n");
	printf("  -i		ipc namespace\n");
	printf("  -P <pid-file>	File in which to write global pid of cinit\n");
	printf("  -p		pid namespace\n");
	printf("  -f <flag>	extra clone flags\n");
	printf("  -e		return child's error status\n");
	printf("  -W		wait to exec the child\n");
    printf("  -s <uid>  setuid to userid\n");
    printf("  -S <uid>  setgid to userid\n");
	printf("\n");
	printf("(C) Copyright IBM Corp. 2006\n");
	printf("\n");
	exit(1);
}

#if 0
static void print_my_info(const char *procname, char *ttyname)
{
	printf("procname %s, ttyname %s, pid %d, ppid %d, pgid %d, sid %d\n",
			procname, ttyname, getpid(), getppid(), getpgid(0),
			getsid(0));
}
#endif

static int string_to_ul(const char *str, unsigned long int *res)
{
	char *tail;
	long long int r;

	if (!*str)
		return -1;

	errno = 0;

	r = strtol(str, &tail, 16);

	/*
	 * according to strtol(3), if errno is set or tail does no point
	 * to the ending '\0', the conversion failed.
	 */
	if (errno || *tail)
		return -1;

	*res = r;
	return 0;
}

/*
 * Copied following opentty() from Fedora's util-linux rpm
 * I just changed the "FATAL" message below from syslog()
 * to printf
 */
static void
opentty(const char * tty) {
        int i, fd, flags;

        fd = open(tty, O_RDWR | O_NONBLOCK);
        if (fd == -1) {
		printf("FATAL: can't reopen tty: %s", strerror(errno));
                sleep(1);
                exit(1);
        }

        flags = fcntl(fd, F_GETFL);
        flags &= ~O_NONBLOCK;
        fcntl(fd, F_SETFL, flags);

        for (i = 0; i < fd; i++)
                close(i);
        for (i = 0; i < 3; i++)
                if (fd != i)
                        dup2(fd, i);
        if (fd >= 3)
                close(fd);
}
// Code copy end

int do_newcgrp = 0;

int load_cgroup_dir(char *dest, int len)
{
	FILE *f = fopen("/proc/mounts", "r");
	char buf[200];
	char *path, *fsname, *options, *p1, *p2, *s;
	if (!f)
		return 0;
	while (fgets(buf, 200, f)) {
		path = strtok_r(buf, " ", &p1);
		path = strtok_r(NULL, " ", &p1);
		fsname = strtok_r(NULL, " ", &p1);
		options = strtok_r(NULL, " ", &p1);
		if (strcmp(fsname, "cgroup") != 0)
			continue;

		/* make sure the freezer is composed */
		s = strtok_r(options, ",", &p2);
		while (s && strcmp(s, "freezer") != 0)
			s = strtok_r(NULL, ",", &p2);
		if (!s)
			continue;
		strncpy(dest, path, len);
		fclose(f);
		return 1;
	}
	fclose(f);
	printf("Freezer not mounted\n");
	return 0;
}

int move_to_new_cgroup(int newcgroup)
{
	char cgroupname[150], cgroupbase[100], tasksfname[200];
	FILE *fout;
	int ret;

	if (!load_cgroup_dir(cgroupbase, 100))
		return 0;

	snprintf(cgroupname, 150, "%s/%d", cgroupbase, newcgroup);
	ret = mkdir(cgroupname, 0755);
	if (ret && errno != EEXIST)
		return 0;
	snprintf(tasksfname, 200, "%s/tasks", cgroupname);
	fout = fopen(tasksfname, "w");
	if (!fout)
		return 0;
	fprintf(fout, "%d\n", getpid());
	fclose(fout);
	return 1;
}

int pipefd[2];

/* gah. opentty will close the pipefd */
int check_newcgrp(void)
{
	int ret, newgroup;
	char buf[20];

	if (!do_newcgrp)
		return 0;

	close(pipefd[1]);
	ret = read(pipefd[0], buf, 20);
	close(pipefd[0]);
	if (ret == -1) {
		perror("read");
		return 1;
	}
	newgroup = atoi(buf);
	if (!move_to_new_cgroup(newgroup))
		return 1;
	do_newcgrp = 0;
	return 0;
}

int wait_for_exec;
int newuid = -1, newgid = -1;

int do_child(void *vargv)
{
	char **argv = (char **)vargv;

	if (check_newcgrp())
		return 1;

	if (wait_for_exec) {
		printf("Press any key to exec (I am %d)\n", getpid());
		(void)getchar();
	}

	if (newgid != -1) {
		if (setgid(newgid) < 0) {
			perror("setgid");
			exit(1);
		}
		if (setgroups(0, NULL) < 0) {
			perror("setgroups");
			return -1;
		}
	}
	if (newuid != -1) {
		if (setuid(newuid) < 0) {
			perror("setuid");
			exit(1);
		}
	}

	execve(argv[0], argv, __environ);
	perror("execve");
	return 1;
}

void write_pid(char *pid_file, int pid)
{
	FILE *fp;

	if (!pid_file)
		return;

	fp = fopen(pid_file, "w");
	if (!fp) {
		perror("fopen, pid_file");
		exit(1);
	}
	fprintf(fp, "%d", pid);
	fflush(fp);
	fclose(fp);
}

int main(int argc, char *argv[])
{	
	int c;
	unsigned long flags = 0, eflags = 0;
	char ttyname[256];
	int status;
	int ret, use_clone = 0, ret_child_err = 0;
	int pid;
	char *pid_file = NULL;
    char *default_args[] = {"/bin/sh", NULL};

	procname = basename(argv[0]);

	memset(ttyname, '\0', sizeof(ttyname));
	readlink("/proc/self/fd/0", ttyname, sizeof(ttyname));

	while ((c = getopt(argc, argv, "+meguUiphcnf:P:Ws:S:")) != EOF) {
		switch (c) {
		case 'g': do_newcgrp = getpid();		break;
		case 'm': flags |= CLONE_NEWNS;			break;
		case 'c': use_clone = 1;			break;
		case 'e': ret_child_err = 1;			break;
		case 'P': pid_file = optarg; 			break;
		case 'u': flags |= CLONE_NEWUTS;		break;
		case 'i': flags |= CLONE_NEWIPC;		break;
		case 'U': flags |= CLONE_NEWUSER;		break;
		case 'n': flags |= CLONE_NEWNET;		break;
		case 'p': flags |= CLONE_NEWNS|CLONE_NEWPID;	break;
		case 'W': wait_for_exec = 1;            break;
		case 's': newuid = atoi(optarg);        break;
		case 'S': newgid = atoi(optarg);        break;
		case 'f': if (!string_to_ul(optarg, &eflags)) {
				flags |= eflags;
				break;
			}
		case 'h':
		default:
			usage(procname);
		}
	};

	argv = &argv[optind];
	argc = argc - optind;	
    if (argc < 1) {
        argv = default_args;
        argc = 1;
    }
	
	if (do_newcgrp) {
		ret = pipe(pipefd);
		if (ret) {
			perror("pipe");
			return -1;
		}
		do_newcgrp = pipefd[0];
	}

    if ((flags & CLONE_NEWUSER) && !wait_for_exec) {
        printf("It is recommended to use wait_for_exec with -U.  Once you\n");
        printf("exec, you lose privileges in the new user namespace and can\n");
        printf("no longer su.\n");
    }

	if (use_clone) {
		int stacksize = 4*getpagesize();
		void *childstack, *stack = malloc(stacksize);

        if (wait_for_exec) {
            printf("Can't use wait_for_exec with clone.  Ignoring it\n");
        }
		if (!stack) {
			perror("malloc");
			return -1;
		}
		childstack = stack + stacksize - 1;

		printf("about to clone with %lx\n", flags);
		flags |= SIGCHLD;
		pid = clone(do_child, childstack, flags, (void *)argv);
		if (pid == -1) {
			perror("clone");
			return -1;
		}
	} else {
		if ((pid = fork()) == 0) {
			// Child.
			//print_my_info(procname, ttyname);

			if (check_newcgrp())
				return 1;
			opentty(ttyname);

			printf("about to unshare with %lx\n", flags);
			ret = unshare(flags);
			if (ret < 0) {
				perror("unshare");
				return 1;
			}		
			
			return do_child((void*)argv);
		}

	}
	if (pid != -1 && do_newcgrp) {
		char buf[20];
		snprintf(buf, 20, "%d", pid);
		close(pipefd[0]);
		write(pipefd[1], buf, strlen(buf)+1);
		close(pipefd[1]);
	}

	write_pid(pid_file, pid);

	if ((ret = waitpid(pid, &status, __WALL)) < 0)
		printf("waitpid() returns %d, errno %d\n", ret, errno);

	if (ret_child_err)
		exit(WEXITSTATUS(status));
	exit(0);
}
